<div style="
    background-color: #7BAFD4;
    border-radius: 18px;
    padding: 22px 26px;
    display: flex;
    align-items: center;
    gap: 24px;
    margin-bottom: 18px;
">
    <img src="592_avatar_lr.png" alt="GEOG 592"
        style="width:110px;height:110px;border-radius:22px;">
    <div>
        <div style="
            font-family: Impact, 'Arial Black', Arial, sans-serif;
            font-size: 42px;
            font-weight: 900;
            letter-spacing: 2px;
            line-height: 1;
            color: white;
        ">GEOG 592</div>
        <div style="
            font-family: Arial, Helvetica, sans-serif;
            font-size: 22px;
            font-weight: 700;
            color: #13294B;
            margin-top: 8px;
        ">GIS Programming</div>
        <div style="
            font-family: Arial, Helvetica, sans-serif;
            font-size: 16px;
            color: #13294B;
            margin-top: 6px;
        ">Homework 3: CSV Cleaning</div>
    </div>
</div>

In [7]:
import csv

filename = "data/NC_Fire_Stations.csv"

with open(filename, "r") as file:
    reader = csv.DictReader(file)
    rows = list(reader)

print("Number of rows:", len(rows))
print("Number of columns:", len(reader.fieldnames))
print("Column names:")
print(reader.fieldnames)

Number of rows: 2126
Number of columns: 13
Column names:
['OBJECTID', 'FD_ID', 'DEPT_NAME', 'STATION_NUMBER', 'STATION_ADDRESS', 'CITY', 'STATE', 'COUNTY', 'ZIP_CODE', 'LATITUDE', 'LONGITUDE', 'x', 'y']


Importing and exploring the structure of the CSV file I found. The file appears to contain 2126 rows, one for each fire station in North Carolina, along with corresponding information like the address and longitude and latitude (x and y coordinates). Therefore, when we clean the data, we will need to keep the x and y coordinates. 

In [ ]:
station_data_cleaned = []

def can_be_float(value):
    try:
        float(value)
        return True
    except (ValueError, TypeError):
        return False

def can_be_int(value):
    try:
        int(value)
        return True
    except ValueError:
        return False


for row in rows:
    xvalue = row["x"]
    yvalue = row["y"]
    
    if can_be_float(xvalue) and can_be_float(yvalue):
        station_data_cleaned.append({
                "name": row["DEPT_NAME"],
                "x": row["x"],
                "y": row["y"]
            })
    else:
         print(f"Error found in coordinates at {row}")
    

Error found in coordinates at {'OBJECTID': '32569', 'FD_ID': '03423', 'DEPT_NAME': 'Winston Salem Fire Department', 'STATION_NUMBER': '21', 'STATION_ADDRESS': '725 North Cherry Street', 'CITY': 'Winston Salem', 'STATE': 'NC', 'COUNTY': 'Forsyth', 'ZIP_CODE': '27102', 'LATITUDE': '', 'LONGITUDE': '', 'x': None, 'y': None}


This code was used to check the x and y coordinates that would be used to spatially join this data. I used a for loop to find any longitude or latitude coordinates that were not able to be converted to a float, meaning they either were missing data or had some other error. The code found that the coordinates are missing for one fire station in Winston Salem. As both the x and y, as well as the latitude and longitude are missing, we have no data to substitute in, so we will just leave it out of the cleaned data output. I appended all float values to a new list along with their name. 

In [ ]:
error_found = False

for row in station_data_cleaned:
    value = row["name"]
    if value == "":
        print(f" {row} is missing a name")
        error_found = True

if not error_found:
    print("Everything looks good.")
    

Everything looks good.


This code chunk is inspecting if every row of coordinates has a name associated with it. As no errors were found, we can concluded that the name column is good!

In [20]:
fieldnames = station_data_cleaned[0].keys() 

with open("data/station_data_cleaned.csv", "w", newline="") as file: 
    writer = csv.DictWriter(file, fieldnames=fieldnames)
    writer.writeheader()
    writer.writerows(station_data_cleaned)

This last bit of code exports the cleaned csv file to the desired data folder. 